deployer/resource_allocation/generate_choices.py

```diff
@@ -0,0 +1,128 @@
+import json
+import sys
+from enum import Enum
+from pathlib import Path
+
+import typer
+from ruamel.yaml import YAML
+
+from ..cli_app import app
+
+yaml = YAML(typ="rt")
+
+HERE = Path(__file__).parent
+
+
+class ResourceAllocationStrategies(str, Enum):
+    PROPORTIONAL_MEMORY_STRATEGY = "proportional-memory-strategy"
+
+
+def proportional_memory_strategy(
+    instance_type: str, nodeinfo: dict, num_allocations: int
+):
+    """
+    Generate choices for resource allocation based on proportional changes to memory
+
+    Used primarily in research cases where:
+    1. Workloads are more memory constrained than CPU constrained
+    2. End users can be expected to select appropriate amount of memory they need for a given
+       workload, either by their own intrinsic knowledge or instructed by an instructor.
+
+    It features:
+    1. No memory overcommit at all, as end users are expected to ask for as much memory as
+       they need.
+    2. CPU *guarantees* are proportional to amount of memory guarantee - the more memory you
+       ask for, the more CPU you are guaranteed. This allows end users to pick resources purely
+       based on memory only, simplifying the mental model. Also allows for maximum packing of
+       user pods onto a node, as we will *not* run out of CPU on a node before running out of
+       memory.
+    3. No CPU limits at all, as CPU is a more flexible resource. The CPU guarantee will ensure
+       that users will not be starved of CPU.
+    4. Each choice the user can make approximately has half as many resources as the next largest
+       choice, with the largest being a full node. This offers a decent compromise - if you pick
+       the largest option, you will most likely have to wait for a full node spawn, while smaller
+       options are much more likely to be shared.
+    """
+
+    # We operate on *available* memory, which already accounts for system components (like kubelet & systemd)
+    # as well as daemonsets we run on every node. This represents the resources that are available
+    # for user pods.
+    available_node_mem = nodeinfo["available"]["memory"]
+    available_node_cpu = nodeinfo["available"]["cpu"]
```

Comment:
This script is operating on available memory/cpu, but there are many reasons for why adjusting to this is tricky:

1. Instance types
   They are relatively easy to adjust to I think
2. The managed k8s cluster's needs changing
   The daemonset's running on each node as managed by the k8s service may vary depending on features enabled (config connector, network policy enforcement, logging), k8s version, and vertical autoscaling determined needs.
3. Our needs changing
   We could end up wanting to add more CPU/RAM requests to the cryptnono daemonset for example.

With that in mind, I think we shouldn't try to be so accurate - because if we are, there is no buffer to still manage to schedule a user requesting 100%.

I suggest we look at the following:
- the capacity of RAM/CPU exposed to k8s pods for various machine types
- the overhead from system pods in GKE, EKS, and AKS respectively - looking at several clusters of each provider to capture variations between enabled features and possibly also k8s versions
- the overhead from our support chart's pods

From this, we take the most conservative available value for every (machine type, cloud provider) combination. Then finally, we remove some capacity to make room for changes over time. I suggest we at least make room for 100m CPU and maybe 250MB RAM or so.

Reply:
I agree we should add some overhead here. I'll incorporate that. However we should still try to be as accurate as possible - you will note that the code that measures how much is available is taking into consideration all the other factors you listed, including the pods that are run as part of our support charts, and whatever it is that the clusters themselves run. I'll actually work on making this even more automated. The overhead is to allow for drift here, as this can change without us noticing - I'll try work on figuring out how best we can 'notice' and correct.

Reply:
@consideRatio https://github.com/2i2c-org/infrastructure/pull/3030/commits/62644fe678cd530bd9b7da6e253d9b4fd05bba66 adds some more flexibility, but much lower than what you recommended. This is because the script *already* accounts for the three things you have listed as needing headroom. This flexibility is now purely to account for changes that we *miss*. I'll work on listing when update-nodeinfo needs to be run, have some inline comments in there.